File: sdks/go/pty.go
package client

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/url"
	"strconv"
	"strings"

	"github.com/coder/websocket"

	"github.com/yanurag-dev/sandforge/pkg/agentproto"
)

// PTYOptions configures an interactive PTY session.
type PTYOptions struct {
	Cols    uint16   // initial terminal width (default 80)
	Rows    uint16   // initial terminal height (default 24)
	Command []string // command to run (default: the guest's login shell)
}

// PTYSession is a client-side handle to an interactive terminal session. It
// mirrors the server-side session contract: NextEvent returns the {event:"exit"}
// event normally, then io.EOF once the session ends.
//
// Concurrency: call SendStdin/Resize from one goroutine and NextEvent from
// another (the WebSocket allows one concurrent reader + one writer). Do not call
// NextEvent concurrently with itself, nor SendStdin concurrently with itself.
type PTYSession struct {
	conn *websocket.Conn
	ctx  context.Context
}

// OpenPTY dials the control plane's PTY WebSocket for the given sandbox and
// starts an interactive session. The ctx governs the whole session lifetime —
// cancelling it tears the session down.
func (c *Client) OpenPTY(ctx context.Context, id string, opts PTYOptions) (*PTYSession, error) {
	wsURL, err := toWebSocketURL(c.baseURL)
	if err != nil {
		return nil, err
	}

	q := url.Values{}
	if opts.Cols > 0 {
		q.Set("cols", strconv.Itoa(int(opts.Cols)))
	}
	if opts.Rows > 0 {
		q.Set("rows", strconv.Itoa(int(opts.Rows)))
	}
	for _, c := range opts.Command {
		q.Add("cmd", c)
	}

	endpoint := wsURL + "/v1/sandboxes/" + id + "/pty"
	if encoded := q.Encode(); encoded != "" {
		endpoint += "?" + encoded
	}

	conn, _, err := websocket.Dial(ctx, endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("open pty: %w", err)
	}
	return &PTYSession{conn: conn, ctx: ctx}, nil
}

// SendStdin forwards input bytes to the remote terminal.
func (s *PTYSession) SendStdin(data []byte) error {
	return s.writeEvent(agentproto.StreamEvent{Event: "stdin", Data: data})
}

// Resize updates the remote terminal window size.
func (s *PTYSession) Resize(cols, rows uint16) error {
	return s.writeEvent(agentproto.StreamEvent{Event: "resize", Cols: cols, Rows: rows})
}

// NextEvent blocks until the next event arrives from the session (stdout/exit/
// error), returning io.EOF once the session has ended. It is a blocking pull,
// not a busy-poll: it parks on the socket read until data arrives.
func (s *PTYSession) NextEvent() (agentproto.StreamEvent, error) {
	_, data, err := s.conn.Read(s.ctx)
	if err != nil {
		return agentproto.StreamEvent{}, normalizeReadErr(err)
	}
	var ev agentproto.StreamEvent
	if err := json.Unmarshal(data, &ev); err != nil {
		return agentproto.StreamEvent{}, fmt.Errorf("decode event: %w", err)
	}
	return ev, nil
}

// Close ends the session and releases the connection.
func (s *PTYSession) Close() error {
	return s.conn.Close(websocket.StatusNormalClosure, "client closed")
}

func (s *PTYSession) writeEvent(ev agentproto.StreamEvent) error {
	data, err := json.Marshal(ev)
	if err != nil {
		return fmt.Errorf("encode event: %w", err)
	}
	return s.conn.Write(s.ctx, websocket.MessageText, data)
}

// normalizeReadErr translates a clean WebSocket close into io.EOF so callers can
// use the idiomatic errors.Is(err, io.EOF) end-of-stream check, matching the
// server-side PTYSession contract. Other errors pass through unchanged.
func normalizeReadErr(err error) error {
	var ce websocket.CloseError
	if errors.As(err, &ce) && ce.Code == websocket.StatusNormalClosure {
		return io.EOF
	}
	return err
}

// toWebSocketURL converts an http(s):// base URL to its ws(s):// equivalent.
func toWebSocketURL(base string) (string, error) {
	switch {
	case strings.HasPrefix(base, "https://"):
		return "wss://" + strings.TrimPrefix(base, "https://"), nil
	case strings.HasPrefix(base, "http://"):
		return "ws://" + strings.TrimPrefix(base, "http://"), nil
	case strings.HasPrefix(base, "ws://"), strings.HasPrefix(base, "wss://"):
		return base, nil
	default:
		return "", fmt.Errorf("unsupported base URL scheme: %q", base)
	}
}

File: sdks/typescript/src/pty.ts
/**
 * sandforge-sdk interactive PTY sessions.
 *
 * A PTY session is a long-lived, full-duplex WebSocket to the control plane.
 * It uses the platform-native WebSocket (browser, or Node 22+ global) so the
 * SDK stays dependency-free.
 *
 * Example:
 *   const session = sandbox.pty.connect({ cols: 120, rows: 40 });
 *   session.send("ls\n");
 *   for await (const ev of session) {
 *     if (ev.event === "stdout") process.stdout.write(Buffer.from(ev.data ?? []));
 *     else if (ev.event === "exit") console.log("exit", ev.code);
 *   }
 */

/**
 * PTYEvent is a single event from a session. `data` carries raw terminal bytes.
 */
export interface PTYEvent {
  event: "stdout" | "exit" | "error";
  data?: Uint8Array;
  code?: number;
  msg?: string;
}

export interface PTYConnectOptions {
  cols?: number;
  rows?: number;
  command?: string[];
}

/**
 * PTYSession is a live interactive terminal session.
 *
 * It is an async-iterable: `for await (const ev of session)` blocks until each
 * event arrives and completes when the session ends — the JS equivalent of the
 * NextEvent()/io.EOF contract used by the other SDKs.
 */
export class PTYSession implements AsyncIterable<PTYEvent> {
  private ws: WebSocket;
  // Bridge the event-based WebSocket to a pull-based async iterator: incoming
  // messages either resolve a waiting next() or buffer in the queue.
  private queue: PTYEvent[] = [];
  private pending: ((r: IteratorResult<PTYEvent>) => void) | null = null;
  private done = false;
  private opened: Promise<void>;

  constructor(url: string, wsImpl?: typeof WebSocket) {
    const WS = wsImpl ?? (globalThis as { WebSocket?: typeof WebSocket }).WebSocket;
    if (!WS) {
      throw new Error(
        "No WebSocket implementation available. Use Node 22+, a browser, or pass one to connect().",
      );
    }
    this.ws = new WS(url);
    this.ws.binaryType = "arraybuffer";

    this.opened = new Promise<void>((resolve, reject) => {
      this.ws.onopen = () => resolve();
      this.ws.onerror = () => reject(new Error("pty websocket error"));
    });

    this.ws.onmessage = (ev: MessageEvent) => this.push(decodeEvent(ev.data));
    this.ws.onclose = () => this.finish();
  }

  /** send forwards input (string or bytes) to the terminal. */
  send(data: string | Uint8Array): void {
    const bytes = typeof data === "string" ? new TextEncoder().encode(data) : data;
    this.ws.send(JSON.stringify({ event: "stdin", data: toBase64(bytes) }));
  }

  /** resize updates the terminal window size. */
  resize(cols: number, rows: number): void {
    this.ws.send(JSON.stringify({ event: "resize", cols, rows }));
  }

  /** close ends the session. */
  close(): void {
    this.ws.close();
    this.finish();
  }

  [Symbol.asyncIterator](): AsyncIterator<PTYEvent> {
    return {
      next: (): Promise<IteratorResult<PTYEvent>> => {
        if (this.queue.length > 0) {
          return Promise.resolve({ value: this.queue.shift()!, done: false });
        }
        if (this.done) {
          return Promise.resolve({ value: undefined, done: true });
        }
        return new Promise((resolve) => {
          this.pending = resolve;
        });
      },
    };
  }

  private push(ev: PTYEvent): void {
    if (this.pending) {
      const resolve = this.pending;
      this.pending = null;
      resolve({ value: ev, done: false });
    } else {
      this.queue.push(ev);
    }
  }

  private finish(): void {
    if (this.done) return;
    this.done = true;
    if (this.pending) {
      const resolve = this.pending;
      this.pending = null;
      resolve({ value: undefined, done: true });
    }
  }
}

/**
 * PTYNamespace opens interactive PTY sessions for a sandbox.
 */
export class PTYNamespace {
  constructor(
    private sandboxID: string,
    private baseURL: string,
    private wsImpl?: typeof WebSocket,
  ) {}

  /** connect opens an interactive PTY session over a WebSocket. */
  connect(opts?: PTYConnectOptions): PTYSession {
    const wsBase = this.baseURL.replace(/^http/, "ws");
    const params = new URLSearchParams();
    if (opts?.cols) params.set("cols", String(opts.cols));
    if (opts?.rows) params.set("rows", String(opts.rows));
    for (const c of opts?.command ?? []) params.append("cmd", c);
    const qs = params.toString();
    const url = `${wsBase}/v1/sandboxes/${this.sandboxID}/pty${qs ? "?" + qs : ""}`;
    return new PTYSession(url, this.wsImpl);
  }
}

function toBase64(bytes: Uint8Array): string {
  if (typeof Buffer !== "undefined") {
    return Buffer.from(bytes).toString("base64");
  }
  let bin = "";
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin);
}

function fromBase64(s: string): Uint8Array {
  if (typeof Buffer !== "undefined") {
    return new Uint8Array(Buffer.from(s, "base64"));
  }
  const bin = atob(s);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

function decodeEvent(raw: unknown): PTYEvent {
  const text =
    typeof raw === "string"
      ? raw
      : new TextDecoder().decode(raw as ArrayBuffer);
  const obj = JSON.parse(text) as {
    event: PTYEvent["event"];
    data?: string;
    code?: number;
    msg?: string;
  };
  return {
    event: obj.event,
    data: obj.data ? fromBase64(obj.data) : undefined,
    code: obj.code,
    msg: obj.msg,
  };
}
